### PR TITLE
[Feature] Paginate by _id

### PR DESCRIPTION
**Pull request by lirbank**
This PR adds two new params `before` and `after`.

These are used just as `next` and `previous`, with the difference that they take a document `_id` instead of a hashed version of `_id` + `paginationFieldValue`, like `next` and `previous` do.

This allows for pagination by _id even when the pagination field is not the _id. It comes at the price of an extra DB lookup, but allows for a shorter page cursor.

I’ll be updating the docs and unit tests but I first wanted to check your stance to merging this feature.

Thoughts?

NOTE: This PR was branched from my other PR so that change is also included in the diff (at the bottom of the file). I'll get the other PR ready to be merged before this one, to remedy that.

### CODE REVIEW

**Comment by bradvogel** [comment]
Thanks! Will review in next 2d

**Comment by lirbank** [comment]
I guess it's a matter of personal preference. Personally, I like the idea of using the _id as the page cursor, even at the cost of an extra lookup. We are using it this way in a GraphQL API for a pretty big app. It enables us to simply pass any document _id as the `after` param in the API and it returns X items after that _id, no special handling or generation/parsing of a "custom cursor", so to speak.

I see what you mean, but in a sense I feel using the _id is actually even simpler (though admittedly more expensive on the DB server), it's an alternative to using next/previous, which does not require a separate cursor (you always have the _id, and you can just pass it to be used as a cursor).

Possibly we should check if more than one of next, previous, after, before are passed, to make it clear to any developer that you can only use one of them at a time?

**Comment by lirbank** [comment]
To clarify - to support paging with sorting by a non-unique field we need 

1) the _id and 
2) the value from the primary sort field.

(you know this, I am just setting the stage)

With `next` both are provided in an encoded string
With `after` only _id is provided, and the primary sort field value is fetched from the DB (via the _id)

From here both ways share the exact same logic.

**Comment by bradvogel** [comment]
I'm still curious on the use case though - how would you sort by something other than _id, but page by _id?

**Comment by lirbank** [comment]
By retrieving the paginatedField value of the _id document, see this line:
https://github.com/mixmaxhq/mongo-cursor-pagination/pull/25/files#diff-b1b9bf02ff07ad669cc803729b4bd013R56

**Comment by lirbank** [comment]
The pagination query logic is the same, I've just added the ability to use another cursor style. 

Instead of a cursor based on encoding _id + paginatedField (`next`), you have the option to use the _id directly as the cursor (`after`).

**Comment by bradvogel** [comment]
Sorry I'm not understanding this. I feel it's breaking the abstraction of the 'after' cursor being an opaque string. This PR makes it so you can construct your own 'after' cursor (by using an _id) with something other than what the library has given you. I can't think of why you wouldn't just do the query yourself using the `{ _id: { $lt: } }`

**Comment by lirbank** [comment]
Hmm, I am not explaining this very well. Let me try again. Use case, let's say you wan't to create an API similar to Stripe's API. They implement cursor based pagination and use the object ID as the cursor, see https://stripe.com/docs/api#pagination - here is a few examples code snippets in different languages (scroll down to find Node.js): http://adamjstevenson.com/stripe/2015/11/22/understanding-pagination-with-the-stripe-api-by-example.html 

I have done something similar but in GraphQL instead of REST. I like the Stripe approach, it's simple and easy to use from an external API user's perspective.

Honest question, why is it desirable to require the cursor to be opaque (asking out of lack of knowledge)? Curious to the rational behind it. I know I read somewhere it should be an opaque string, but never understand why/if that is important. To me an _id would be just as good, but I may be missing something here (totally possible Inc. :)). Here is some info around it http://graphql.org/learn/pagination/ - among the reasons they give are "The ability to change how our backend does pagination, since the user just uses opaque cursors". But here I think they talk about modify the GraphQL resolvers, not replacing the actual DB, I think (not sure).

Actually, are the current cursors of this package really truly opaque? I mean since they are last_id_value+last_sort_field_value combined into a string, they are kind of semi-transparent too... Not sure how to think about this.

Why not just use `{ _id: { $lt: } }`? This will not allow me to sort by any other field than the _id field. If you look at my contribution, it will allow me to use all the features of your package while using any _id as the cursor instead of the cursor string returned from the lib. So all features, except the generated cursor string, could be used (has more, sorting field and sorting direction, etc). I've tried it, and it works as far as I can see.

Actually, I could check if the inputs are before/after, and replace the returned cursor with the last _id, to make it more compatible (right now it is a bit confusing that the lib returns a cursor that you should not use, but this was all I needed to get it to work for us, and I wanted to ask if you'd be OK with merging the feature before doing the polishing).

**Comment by bradvogel** [comment]
Sorry for the delay. I think I understand this a bit more after reading the Stripe docs.

Stripe's API is different because they don't support sorting. Once you add in sorting, the cursor string _must_ be opaque.

Take, for example, the following MongoDB records:
{ _id: 'a', value: 2 }
{ _id: 'b', value: 3 }
{ _id: 'c', value: 1 }
{ _id: 'd', value: 4 }

In the Stripe API case, everything is sorted by the `_id`, so fetching the first page by `_id` is this simple:
`find({}).sort({ _id: -1 }).limit(2); // returns [{value: 2},{value: 3}]`

And the second page:
`find({ _id: { $lt: 'b' } }).sort({ _id: -1 }).limit(2); // returns [{value: 1},{value: 4}]`

However, once you sort by something other than the `_id`, say `value` in this case, the `_id` because meaningless. So the first page would be fetched by:
`find({}).sort({ value: -1 }).limit(2); // returns [{value: 4},{value: 3}]`

And the second page:
`find({ value: { $lt: 3 } }).sort({ value: -1 }).limit(2); // returns [{value: 2},{value: 1}]`

So if Stripe were to add a `sort` querystring parameter to their API, they would need to change their pagination docs from: "just pass the last objectid back to us in `starting_after` parameter" to "if you're sorting by field X, pass the last value of field X back as the `starting_after` parameter" (even that won't work because some values might have duplicates that require to use guaranteed `_id` as a tiebreaker).

So the thing I don't like about this feature is that 'after' and 'before' only work if you're NOT using the sort feature (paginatedField), which kind of defeats the purpose of the simple API.

If you wanted to expose restrict the query to only query a range of _ids, you could also pass it in the query, e.g. `.find({ query: { _id: { $lt: _id } } })`

Perhaps this module should be called mongo-cursor-pagination-with-sorting :).

**Comment by lirbank** [comment]
This PR does _not_ restrict you to sorting by the ID, you can still sort by any field, without missing documents.

Can you take another look at the code change and I think you will see what I mean?

**Comment by lirbank** [comment]
If what I say is true, would you merge? If so I could just write the tests to prove it.

The PR is not ready to be merged anyhow, I first wanted to know if you'd accept the feature before spending the time finishing it up.

**Comment by bradvogel** [comment]
Ok, please write a test that demonstrates the value of using the API in this way:

```
find({
  limit: 2,
  paginatedField: 'timestamp',
  next: '<cursor>',
  after: '<id>'
})
```

**Comment by lirbank** [comment]
I've added all test you had for next/previous to also test after/before. I also fixed an issue where after/before did not like the paginatedField using dot notation. All tests pass.

To be clear, you should _not_ use `after`/`before` together with `next`/`previous`. Just like you should not use `previous` and `next` at the same time. I mean you can, but only one of them will be used. 

Currently `next` overrides `previous` (not my code). Then `after` overrides `next`, and `before` overrides `previous` (my code). In short, only one should be used at a time. This could be clarified with a warning or by throwing an error if more than one of next, previous, before or after is provided.

As you can see in the tests, you use an _id from a previous result to move the cursor. In the tests I take the first _id (for `before`) or the last _id (for `next`), to replicate the behavior of the other tests (moving full pages). But you could just as well use an _id in the middle of the result to move half a page.

It's really flexible; it's even possible to use `next` (using opaque string) for the second page and then load the third page with `after` (_id), and so on. Not saying anyone would do that (or should), but they are compatible so it is possible.

**Comment by bradvogel** [comment]
Ok merged. Published as 6.2. Thanks!
